hotkey: Support non-Latin keyboard locales for shortcuts.

prakhar1144 opened the pull request:
* PR to support non-Latin keyboard locales for shortcuts. 
* Includes a commit to fix Caps Lock incorrectly affecting Shift state in shortcuts. -- [#issues > Shortcuts in non-latin languages #19605 @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/Shortcuts.20in.20non-latin.20languages.20.2319605/near/2171089)

Fixes: https://github.com/zulip/zulip/issues/19605

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

gnprice commented on web/src/hotkey.js:
```diff
@@ -1316,31 +1418,17 @@ export function process_hotkey(e, hotkey) {
 
-/* We register both a keydown and a keypress function because
-   we want to intercept PgUp/PgDn, Esc, etc, and process them
-   as they happen on the keyboard. However, if we processed
-   letters/numbers in keydown, we wouldn't know what the case of
-   the letters were.
-
-   We want case-sensitive hotkeys (such as in the case of r vs R)
-   so we bail in .keydown if the event is a letter or number and
-   instead just let keypress go for it. */
```
Interesting, what's the story of this comment? It seems to be explaining a reason why `keydown` events won't suffice for some of our needs.

Have you confirmed that the thing it warns about seems to still work fine? Have you been able to work out what changed since this comment was written?

prakhar1144 replied:
> However, if we processed letters/numbers in keydown, we wouldn't know what the case of the letters were.

Let's take `r` vs `R` as an example.

Earlier, we were using `event.which`.

For `r`:
| Event type | which |
| --- | --- |
| keypress | 114 |
| keydown | 82 |

For `R`:
| Event type | which |
| --- | --- |
| keypress | 82 |
| keydown | 82 |

So, if we use `event.which` with `keydown`, the value is `82` for both `r` and `R`. No way to distinguish. So, at previously we were taking help of `keypress` -- `114` vs `82`.

But now we're using `.key` which clearly has the value `r` vs `R` in `keydown`. So, we can safely stop relying on `.keypress`.

prakhar1144 replied:
> Have you confirmed that the thing it warns about seems to still work fine?

I have manually tested few of the hotkeys which were earlier managed by `keypress` and now managed by `keydown`; they work as expected. E.g. `a` vs `A`

> Have you been able to work out what changed since this comment was written?

`.which` is now legacy. `.key` is the thing now supported by major browsers -- https://caniuse.com/?search=KeyboardEvent.key

Btw, the finalized version of the spec was released just ~2 weeks ago (https://www.w3.org/TR/uievents-key/). But majority of the browsers have already implemented the draft specs, so I think we're safe to use it.

timabbott replied:
Let's copy some of this explanation into the commit message for posterity.

gnprice replied:
Great, thanks for that explanation.

It sounds like the direct answer to "what changed since this comment was written" is: we switched from `.which` to `.key`, in one of the earlier commits in this very PR. :-) So drawing that connection will be one key piece of information to put into the commit message.

prakhar1144 replied:
Updated the commit message.

gnprice commented on web/src/hotkey.js:
```diff
@@ -100,52 +100,94 @@ const color_picker_hotkeys = new Set([
+        case "PageUp":
+            return {name: "page_up", message_view_only: true};
+        case "PageDown":
+            return {name: "page_down", message_view_only: true};
+        case "End":
+            return {name: "end", message_view_only: true};
+        case "Home":
+            return {name: "home", message_view_only: true};
+        case "ArrowLeft":
+            return {name: "left_arrow", message_view_only: false};
+        case "ArrowRight":
+            return {name: "right_arrow", message_view_only: false};
+        case "ArrowUp":
+            return {name: "up_arrow", message_view_only: false};
+        case "ArrowDown":
+            return {name: "down_arrow", message_view_only: false};
+        default:
+            return undefined;
+    }
+}
 
-const keydown_ctrl_mappings = {
-    219: {name: "escape", message_view_only: false}, // '['
-};
+function keydown_ctrl_mappings(key) {
+    switch (key) {
+        case "[":
+            return {name: "escape", message_view_only: false};
```
hotkey: Replace deprecated `e.which` with `e.key`.
    
    The `keydown` and `keypress` event handlers defined in `hotkey.js`
    were using `event.which`, which is deprecated.
    
    This commit makes changes to use `event.key` instead.

Interesting. Does this change any of the behavior? I'm not sure how the semantics of `.which` compares to the semantics of `.key`.

Definitely an improvement to have these strings instead of the numbers.

prakhar1144 replied:
I think [the above comment](https://github.com/zulip/zulip/pull/34570#discussion_r2078994171) explains this ?

gnprice replied:
Well, that comment explains an additional reason why we'd like to be using `.key` (beyond "`.which` is old and deprecated").

My question in this subthread, though, is: does the behavior change with this commit, as seen by the user? Is there any situation where a user hits a key (or combination of keys) and the code at this commit's parent would do one thing, but the code at this commit would do a different thing?

If the answer is no, that's helpful for thinking about the commit and understanding it.

If the answer is yes, that's also important to know for thinking about the commit — and the commit message should also help the reader understand what those differences in behavior are. For example, is there a situation where the code was buggy before this, and this commit fixes the bug?

prakhar1144 replied:
Okay.

"No behavioral change for the end user."

Included that in the commit message.

gnprice replied:
Continued below: https://github.com/zulip/zulip/pull/34570#discussion_r2087613005

gnprice commented on web/src/hotkey.js:
```diff
@@ -157,81 +199,144 @@ const keydown_either_mappings = {
     // For Enter, there is some possibly that Shift-Enter is intended to
     // have special behavior for folks that are used to Shift-Enter behavior
     // in other apps, but that's also slightly dubious.
-    8: {name: "backspace", message_view_only: true}, // Backspace
-    13: {name: "enter", message_view_only: false}, // Enter
-    46: {name: "delete", message_view_only: false}, // Delete
-};
+    switch (key) {
+        case "Backspace":
+            return {name: "backspace", message_view_only: true};
+        case "Enter":
+            return {name: "enter", message_view_only: false};
+        case "Delete":
+            return {name: "delete", message_view_only: false};
```
In the last commit, it's unfortunate that the `switch` version makes everything twice as long ­— the list becomes harder to scan by eye. It doesn't seem like at this stage that change brings any benefits, either.

Perhaps it becomes helpful as part of the next round of changes after it? In that case it might be best to save for that next PR.

prakhar1144 replied:
Today, I added the 8th commit in this PR itself. I realised, we should use switch statements in `character_mappings` and `keydown_cmd_or_ctrl_mappings` only.

In other mappings, only one or two such scenarios occurs where multiple keys points to the same value. It's best to not use switch statement in those mappings and just make separate entries duplicating the values. For example:
```js
// Instead of
function keydown_ctrl_mappings(key) {
    switch (key) {
        case "[":
        case "BracketLeft":
            return {name: "escape", message_view_only: false};
        default:
            return undefined;
    }
}

// simply
const keydown_ctrl_mappings = {
    [ :  {name: "escape", message_view_only: false},
    BracketLeft : {name: "escape", message_view_only: false},
}
```

gnprice replied:
Well, if we end up with `switch` for `character_mappings`, then I'd probably want to do the same for the rest too.

The thing is that `character_mappings` is by far the biggest of them — it looks like it's over half the total length all by itself. So at that point there isn't so much to be gained by putting just the other mappings into a more compact format; and OTOH there's a readability advantage from keeping it uniform.

prakhar1144 replied:
Posted a slightly different approach in [#issues > Shortcuts in non-latin languages #19605 @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/Shortcuts.20in.20non-latin.20languages.20.2319605/near/2170650)

prakhar1144 replied:
Implemented the discussed approach. See the last commit.

gnprice reviewed (commented):
Excited to see this issue #19605 progressing!

prakhar1144 commented on web/tests/hotkey.test.cjs:
```diff
@@ -204,8 +197,6 @@ run_test("mappings", () => {
 
     // More negative tests.
-    assert.equal(map_down("/"), undefined);
-    assert.equal(map_press("Escape"), undefined);
```
These two were earlier passing because `"/"` was handled by `keypress` and `"Escape"` by `keydown`. Now, both are handled by `keydown` so just removed it from here as we already test them earlier in this test file.

andersk commented on web/src/hotkey.js:
```diff
@@ -157,81 +208,222 @@ const keydown_either_mappings = {
+    }
+
+    const code_point = key.codePointAt(0);
+    return code_point >= 32 && code_point <= 126;
```
Could simplify:

```suggestion
    return key.length === 1 && key >= " " && key <= "~";
```

andersk commented on web/src/hotkey.js:
```diff
@@ -157,81 +208,222 @@ const keydown_either_mappings = {
+}
+
+function is_printable_ascii(key) {
+    if (key.length !== 1) {
+        return false;
+    }
+
+    const code_point = key.codePointAt(0);
+    return code_point >= 32 && code_point <= 126;
```
Could simplify:

```suggestion
    return key.length === 1 && key >= " " && key <= "~";
```

prakhar1144 replied:
Yeah. Made the changes.

prakhar1144 commented on web/src/hotkey.js:
```diff
@@ -177,15 +174,20 @@ const keypress_mappings = {
+    V: [
+        {name: "view_selected_stream", message_view_only: false},
+        {name: "toggle_read_receipts", message_view_only: true},
+    ],
```
Let's do it as a followup.

prakhar1144 commented:
Updates in the recent push:
* Addressed [#issues > Shortcuts in non-latin languages #19605 @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/Shortcuts.20in.20non-latin.20languages.20.2319605/near/2171089)
* Addressed review comments
* A new commit: 5th commit.

TODO:
* Write node tests for the "non-latin" case. -- I'm wondering if we want to add tests (for non-latin case) in parallel for all the hotkeys present + the various negative cases present in `hotkey.test.cjs`. Or we should test just a handful of them ?

timabbott commented on web/src/hotkey.js:
```diff
@@ -1316,31 +1315,17 @@ export function process_hotkey(e, hotkey) {
     activity.set_new_user_input(true);
-    const hotkey = get_keydown_hotkey(e);
-    if (!hotkey) {
+    const result = get_keydown_hotkey(e);
```
Probably ideally we should be coming up with a name for the type of the objects that `get_keydown_hotkey` returns and referring to that here rather than the generic `result`, which would usually be a good name for something we'll return; we can defer that for when we do the TS migration for this module, I guess.

prakhar1144 replied:
Yeah, I spent a few minutes to figure out a better variable name -- which indicates that it can be a single hotkey or an array. Ended up with a very generic `result`.

timabbott commented on web/src/hotkey.js:
```diff
@@ -213,27 +213,111 @@ const character_mappings = {
 };
 
+function is_printable_ascii(key) {
+    return key.length === 1 && key >= " " && key <= "~";
```
It'd be helpful to have a comment documenting what set of characters this is meant to cover. It's not at all obvious to me that this corresponds exactly to 'printable ascii" characters...

prakhar1144 replied:
Added comment
```
    // ASCII printable characters (character code 32-126) -> " " to "~".
    // It includes letters, digits, punctuation marks, and a few
    // miscellaneous symbols.
```

gnprice commented on web/src/hotkey.js:
```diff
@@ -102,47 +102,47 @@ const color_picker_hotkeys = new Set([
-    37: {name: "left_arrow", message_view_only: false}, // left arrow
-    39: {name: "right_arrow", message_view_only: false}, // right arrow
-    38: {name: "up_arrow", message_view_only: false}, // up arrow
-    40: {name: "down_arrow", message_view_only: false}, // down arrow
+    Tab: {name: "tab", message_view_only: false},
+    Escape: {name: "escape", message_view_only: false},
+    " ": {name: "spacebar", message_view_only: true},
+    PageUp: {name: "page_up", message_view_only: true},
+    PageDown: {name: "page_down", message_view_only: true},
+    End: {name: "end", message_view_only: true},
+    Home: {name: "home", message_view_only: true},
+    ArrowLeft: {name: "left_arrow", message_view_only: false},
+    ArrowRight: {name: "right_arrow", message_view_only: false},
+    ArrowUp: {name: "up_arrow", message_view_only: false},
+    ArrowDown: {name: "down_arrow", message_view_only: false},
 };
 
 const keydown_ctrl_mappings = {
-    219: {name: "escape", message_view_only: false}, // '['
+    "[": {name: "escape", message_view_only: false},
 };
 
 const keydown_cmd_or_ctrl_mappings = {
-    13: {name: "action_with_enter", message_view_only: true}, // 'Enter'
-    67: {name: "copy_with_c", message_view_only: false}, // 'C'
-    75: {name: "search_with_k", message_view_only: false}, // 'K'
-    83: {name: "star_message", message_view_only: true}, // 'S'
-    190: {name: "narrow_to_compose_target", message_view_only: true}, // '.'
-    222: {name: "open_saved_snippet_dropdown", message_view_only: true}, // '''
+    Enter: {name: "action_with_enter", message_view_only: true},
+    c: {name: "copy_with_c", message_view_only: false},
+    k: {name: "search_with_k", message_view_only: false},
+    s: {name: "star_message", message_view_only: true},
+    ".": {name: "narrow_to_compose_target", message_view_only: true},
+    "'": {name: "open_saved_snippet_dropdown", message_view_only: true},
 };
 
 const keydown_alt_mappings = {
-    80: {name: "toggle_compose_preview", message_view_only: true}, // 'P'
+    p: {name: "toggle_compose_preview", message_view_only: true},
```
It's surprising to me that this change would have no effect on the behavior:

    hotkey: Replace deprecated `e.which` with `e.key`.
    
    The `keydown` and `keypress` event handlers defined in `hotkey.js`
    were using `event.which`, which is deprecated.
    
    This commit makes changes to use `event.key` instead.
    
    No behavioral change for the end user.

The semantics of `e.which` seems quite complex, and the semantics of `e.key` aren't simple either. Do they really align so neatly in the way we use them?

Playing with the tool you linked in another thread:
https://w3c.github.io/uievents/tools/key-event-viewer.html
(which BTW would be good to link in a commit message, or maybe a comment in the code), it looks like one important difference in semantics is: for a letter X, Shift+X and just X have the same `e.which`, but different `e.key`.

Then looking at the logic in `get_keydown_hotkey`, I think one difference in user-visible behavior is right here:
* Today in main, Alt+Shift+P invokes the preview shortcut just like Alt+P does.
* After this commit, I believe it won't.

So that's a good change — it fixes a small bug.

We should understand that we're changing something in the behavior, though, and what we're changing. That's important for our ability to notice if there's also something changing that we don't want.

prakhar1144 replied:
> Shift+X and just X have the same e.which in their keydown events, but different e.key.

Yeah, that's the reason we were using `keypress` events earlier to distinguish.

> The semantics of e.which seems quite complex, and the semantics of e.key aren't simple either. Do they really align so neatly in the way we use them?

Both `.which` and `.key` provide information about the key, the user pressed. 

I'm looking it as:
* `.which` was the old key combined with old mappings used to give us the hotkey. 
* `.key` is the new key and we have new mappings.

`.which` is deprecated -- value not compatible across browsers and other reasons.

> Today in main, Alt+Shift+P invokes the preview shortcut just like Alt+P does.

Yeah, Ideally we should fix this in a separate commit as:
```js
    if (e.altKey && !e.shiftKey) {
        hotkey = keydown_alt_mappings[e.which];
        if (hotkey) {
            return hotkey;
        }
        return undefined;
    } else if (e.altKey) {
        return undefined;
    }
```
but we are reworking this codepath to have descriptive keys, so avoiding a separate commit here as the changes would be replaced in a later commit.

`Alt + P` wasn't working for me as discussed in CZO, but yeah one can catch this bug by dry-running the code.

Updated the commit message as this gets fixed as a side effect of this change.

gnprice commented on web/src/hotkey.js:
```diff
@@ -157,81 +195,129 @@ const keydown_either_mappings = {
 export function get_keydown_hotkey(e) {
     let hotkey;
+    const use_event_key = is_printable_ascii(e.key) || NAMED_KEY_ATTRIBUTE_VALUES.has(e.key);
+    let event_key = e.key;
+    if (!use_event_key) {
+        const code = `${e.shiftKey ? "shift-" : ""}${e.code}`;
+        event_key = CODE_TO_QWERTY_CHAR[code];
```
The naming of these locals is confusing. There's a boolean `use_event_key`  — but then even when it's false, we end up using the thing called `event_key`?

I think renaming `event_key` to just `key` would probably fix that problem.

prakhar1144 replied:
> renaming event_key to just key would probably fix that problem.

Done

gnprice commented on web/src/hotkey.js:
```diff
@@ -157,81 +195,129 @@ const keydown_either_mappings = {
+const CODE_TO_QWERTY_CHAR = {
+    BracketLeft: "[",
+    Equal: "=",
+    KeyA: "a",
+    KeyC: "c",
+    KeyD: "d",
+    KeyE: "e",
+    KeyG: "g",
```
This map should be made complete, at least for the letters, rather than have only the letters we happen to have shortcuts for today.

That will save people in the future from some confused debugging when they try adding a shortcut in the table of shortcuts and it doesn't seem to do anything (because it isn't also in this table). Confusing especially because they may have done the exact same thing the previous day for a different shortcut and it *did* work (because that one was already here, because there was an existing shortcut with different modifier keys on the same letter).

gnprice replied:
Probably also most readable if the non-letters are grouped together either before or after the letters, rather than be split based on whether their names happen to come before or after "Key" alphabetically.

prakhar1144 replied:
> This map should be made complete, at least for the letters

Done

> Probably also most readable if the non-letters are grouped together

Done

gnprice commented on web/src/hotkey.js:
```diff
@@ -157,81 +195,129 @@ const keydown_either_mappings = {
+    return key.length === 1 && key >= " " && key <= "~";
+}
+
+const NAMED_KEY_ATTRIBUTE_VALUES = new Set([
```
Is this a complete list of named key attribute values?

Looks like not — there's a long and kind of unbounded list:
https://www.w3.org/TR/uievents-key/#named-key-attribute-values
so it'll inevitably be incomplete and we may need to add to it sometimes.

This should get either a comment, or a name (like starting with "known"), to make clear that it's incomplete and just aims to contain the ones that we use in shortcuts.

prakhar1144 replied:
Renamed it to `KNOWN_NAMED_KEY_ATTRIBUTE_VALUES`

andersk commented on web/src/hotkey.js:
```diff
@@ -241,28 +327,25 @@ export function get_keydown_hotkey(e) {
 
     if (e.shiftKey) {
-        hotkey = keydown_shift_mappings[e.which];
+        hotkey = keydown_shift_mappings[event_key];
         if (hotkey) {
             return hotkey;
         }
     }
 
     if (!e.shiftKey) {
-        hotkey = keydown_unshift_mappings[e.which];
+        hotkey = keydown_unshift_mappings[event_key];
```
Ideally we would use the event to construct a fully descriptive string like `Ctrl+Alt+Shift+A`, and then look up that string in a single table, rather than having all of these branches to decide which one or more of five tables to consult.

prakhar1144 commented on web/src/hotkey.js:
```diff
@@ -241,28 +327,25 @@ export function get_keydown_hotkey(e) {
-    if (e.metaKey || e.ctrlKey || e.altKey) {
-        return undefined;
+    hotkey = keydown_either_mappings[e.key];
+    if (hotkey) {
+        return hotkey;
     }
 
-    return keypress_mappings[e.which];
+    return character_mappings[e.key];
```
For `keydown_either_mappings`, the comment says:
```
// these can be triggered by key or Shift + key
```
which is not true for `character_mappings`.

prakhar1144 replied:
We'll have only one map with this approach : https://github.com/zulip/zulip/pull/34570#discussion_r2087771690, so marking this as resolved.

prakhar1144 commented on web/src/hotkey.js:
```diff
@@ -102,22 +102,57 @@ const color_picker_hotkeys = new Set([
-    ArrowRight: {name: "right_arrow", message_view_only: false},
-    ArrowUp: {name: "up_arrow", message_view_only: false},
-    ArrowDown: {name: "down_arrow", message_view_only: false},
+    A: {name: "stream_cycle_backward", message_view_only: true},
```
> Do these need to be power-cased?

Yes

> I'm not sure if this handles the situations where you hit shift+A with caps lock on correctly.

Yeah, it was working in that case. It was not working in situations where you hit `A` with Caps Lock on.

Anyways, we have reworked it to have a [single map with descriptive keys](https://github.com/zulip/zulip/pull/34570#discussion_r2087771690). No `a` vs `A` now, it'd be `A` vs `Shift+A`.

prakhar1144 commented on web/src/hotkey.js:
```diff
@@ -157,81 +195,129 @@ const keydown_either_mappings = {
+    let event_key = e.key;
+    if (!use_event_key) {
+        const code = `${e.shiftKey ? "shift-" : ""}${e.code}`;
+        event_key = CODE_TO_QWERTY_CHAR[code];
```
Yeah, it was specifically resolving the bug where pressing a letter key, lets say `A`, with caps lock enabled resulted in `"stream_cycle_backward"`. Ideally that should happen only when `shift+A` is used.

So, `shift+key` combinations were strict. Only using `key` with caps lock enabled won't trigger hotkeys that require `shift`.

But, that commit introduced a bug where pressing `a` with caps lock enabled was not working -- which is fixable with a small tweak...

Anyways, I have reworked it with the discussed approach in CZO as it solves one more bug other than the Caps Lock thing (see commit message of 7th commit) -- and i feel this approach of having descriptive keys is going to be helpful in future to avoid bugs.

prakhar1144 commented:
Addressed the review comments.

prakhar1144 commented on web/src/hotkey.js:
```diff
@@ -163,107 +196,49 @@ const keydown_either_mappings = {
+    // We ignore the modifiers for these keys as they are treated as
+    // involved in generating the character in various keyboard layouts.
+    // Here, modifiers are not the part of the shortcut key combination.
+    const ignore_modifiers_for_keys = new Set(["*", "+", "=", "-", "/", ":", "<", ">", "?", "@"]);
```
TODO: We should define this set outside the function.

timabbott replied:
Yeah we can define this just before `KEYDOWN_MAPPINGS`, and call it `IGNORE_MODIFIERS_KEYS` with a comment along the lines of:

> This collection of characters are those where many keyboards require modifier keys in order to produce, or where pressing this key with a modifier key pressed is expected to produce a different key . We ignore all modifiers when processing these keys, and disallow assigning hotkeys to any of these.

prakhar1144 replied:
```js
// A set of characters that, on many keyboard layouts, require modifier keys
// to type, or where pressing the key with a modifier is expected to produce
// a different character. We ignore all modifiers when processing these keys,
// and disallow assigning hotkeys that involve any modifier combined with them.
const KNOWN_IGNORE_MODIFIERS_KEYS = new Set(["*", "+", "=", "-", "/", ":", "<", ">", "?", "@"]);
```
Updated the comment and moved it out of the function.

Also, added a prefix `KNOWN` to the variable name as it is not a complete list of such characters but the ones we use currently. -- Same reason as https://github.com/zulip/zulip/pull/34570#discussion_r2087649881

timabbott commented on web/src/hotkey.js:
```diff
@@ -163,107 +196,49 @@ const keydown_either_mappings = {
+        }
+        if (e.metaKey) {
+            assert(common.has_mac_keyboard());
+            key = "Cmd+" + key;
```
Why is this assertion true? Does the spec say only mac keyboard systems can have a `metaKey`? I suspect this logic is wrong.

timabbott replied:
Maybe we want this to be:
```
if (e.metaKey) {
    if (common.has_mac_keyboard()) {
        key = "MacCmd" + key;
    } else {
        key = "Cmd" + key;
    }
} 
```

and we'd have our CmdOrCtrl legacy keys just use `MacCmd` as the name -- so if you use a `Cmd` type key (I think the Windows key on non-Mac computers?) that doesn't fall into the Mac logic -- on a PC, you want `Ctrl` to do that and `Cmd` to not behave like `Ctrl` unexpectedly.

prakhar1144 replied:
> Why is this assertion true? Does the spec say only mac keyboard systems can have a metaKey? I suspect this logic is wrong.

For non-mac systems, our docs suggest to use `Ctrl` instead of `Cmd`. So, if `Cmd` is involved in a shortcut combination on a non-mac system, shouldn't we just return undefined ?

prakhar1144 replied:
```
    const isCmdOrCtrl = common.has_mac_keyboard() ? e.metaKey : e.ctrlKey;
    if (isCmdOrCtrl && !e.shiftKey) {
        hotkey = keydown_cmd_or_ctrl_mappings[e.key];
        if (hotkey) {
            return hotkey;
        }
        return undefined;
    } else if (e.metaKey || e.ctrlKey) {
        return undefined;
    }
    ```
    Earlier, this was our code. In case of mac, use `Cmd` otherwise `Ctrl`. If we're using `Meta` key on windows, that'd return `undefined`.

prakhar1144 replied:
This also resolves a bug as mentioned in commit message:
> This descriptive string approach also resolves a bug where
`Ctrl + Meta + C/K/S/.` was working. We define the valid hotkey as
`Meta + C/K/S/.` for macOS and `Ctrl + C/K/S/.` for others.

timabbott commented on web/src/hotkey.js:
```diff
@@ -201,10 +201,113 @@ const KEYDOWN_MAPPINGS = {
+    if (!use_event_key) {
+        const code = `${e.shiftKey ? "Shift+" : ""}${e.code}`;
+        key = CODE_TO_QWERTY_CHAR[code];
+    }
```
This little algorithm at the top here needs a nice comment explaining its purpose -- probably there's a nice comment in this PR that we can copy/paste in.

prakhar1144 replied:
Added a comment:
```js
    // Determine the key pressed in a consistent way.
    //
    // For keyboard layouts (e.g. QWERTY) where `e.key` results
    // in printable ASCII characters or named key attribute values
    // like "Tab", "Enter", etc., we use `e.key` directly.
    //
    // However, for layouts where the character defined for a hotkey
    // can't be typed directly (e.g. 'a' in Cyrillic), users press
    // the same physical key combination that a QWERTY user would
    // use (e.g. 'ф' on Cyrillic maps to 'a' on QWERTY). In such cases,
    // we derive the QWERTY equivalent using `e.code` and the `CODE_TO_QWERTY_CHAR` map.
```

prakhar1144 commented:
Addressed the review comments.

prakhar1144 commented:
In the latest push:
* Added node test in the last commit.
* A long commit message for the last commit.

No functional change.

andersk commented on web/src/hotkey.js:
```diff
@@ -89,6 +89,12 @@ const color_picker_hotkeys = new Set([
+// to type, or where pressing the key with a modifier is expected to produce
+// a different character. We ignore all modifiers when processing these keys,
+// and disallow assigning hotkeys that involve any modifier combined with them.
+const KNOWN_IGNORE_MODIFIERS_KEYS = new Set(["*", "+", "=", "-", "/", ":", "<", ">", "?", "@"]);
```
This mechanism is problematic; for example, it causes us to misinterpret `Ctrl`+`+` (browser zoom) as `+` (thumbs up emoji). [Discussion](https://chat.zulip.org/#narrow/channel/9-issues/topic/browser.20zoom.20shortcuts.20not.20working/near/2182431).

timabbott commented:
Merged, thanks @prakhar1144!
